test/jdk/java/util/TimeZone/CustomTzIDCheckDST.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8285838
+ * @library /test/lib
+ * @summary This test will ensure that daylight savings rules are followed
+ * appropriately when setting a custom timezone ID via the TZ env variable.
+ * @requires os.family != "windows"
+ * @run main/othervm CustomTzIDCheckDST
+ */
+import java.util.Calendar;
+import java.util.Date;
+import java.util.List;
+import java.time.Month;
+import java.time.DayOfWeek;
+import java.time.ZonedDateTime;
+import java.time.temporal.TemporalAdjusters;
+import java.time.ZoneId;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+public class CustomTzIDCheckDST {
+    private static String CUSTOM_TZ = "MEZ-1MESZ,M3.5.0,M10.5.0";
+    public static void main(String args[]) throws Throwable {
+        if (args.length == 0) {
+            ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(List.of("CustomTzIDCheckDST", "runTZTest"));
+            pb.environment().put("TZ", CUSTOM_TZ);
+            OutputAnalyzer output = ProcessTools.executeProcess(pb);
+            output.shouldHaveExitValue(0);
+        } else {
+            runTZTest();
+        }
+    }
+
+    /* TZ code will always be set to "MEZ-1MESZ,M3.5.0,M10.5.0".
+     * This ensures the transition periods for Daylights Savings should be at March's last
+     * Sunday and October's last Sunday.
+     */
+    private static void runTZTest() {
+        Calendar calendar = Calendar.getInstance();
+        Date time = calendar.getTime();
+        int month = time.getMonth();
+        ZonedDateTime date = ZonedDateTime.ofInstant(time.toInstant(), ZoneId.systemDefault());
+
+        if ((month > Month.MARCH.getValue() && month < Month.OCTOBER.getValue()) ||
+                (month == Month.MARCH.getValue() && date.isAfter(getLastSundayOfMonth(date))) ||
+                (month == Month.OCTOBER.getValue() && date.isBefore(getLastSundayOfMonth(date)))) {
```

Comment:
I don't think these conditions are correct, as `month` is zero-based, and comparing it with `Month.MARCH` will be incorrect. The same holds for October.

Reply:
Oh thats true, just verified this by printing out the values. Will fix by adding 1 to the month for sensible direct comparison against the singleton instances.